File: week03/Venkatesh Yadav Edamoni WEEK 03.py
import pandas as pd
import numpy as np
import time

df = pd.read_excel("clinics.xls")

def haversine(lat1, lon1, lat2, lon2):
    """Calculate distance between two points in miles"""
    MILES = 3959
    lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a)) 
    return MILES * c

ref_lat = df.iloc[0]['locLat']
ref_long = df.iloc[0]['locLong']

# Method 1: Simple for loop
def method_loop():
    start_time = time.time()
    distances = []
    for i in range(len(df)):
        d = haversine(ref_lat, ref_long, 
                     df.iloc[i]['locLat'], 
                     df.iloc[i]['locLong'])
        distances.append(d)
    end_time = time.time()
    return distances, end_time - start_time

# Method 2: Using pandas apply
def method_apply():
    start_time = time.time()
    distances = df.apply(lambda row: haversine(ref_lat, ref_long,
                                             row['locLat'],
                                             row['locLong']), axis=1)
    end_time = time.time()
    return distances, end_time - start_time

# Method 3: Vectorized approach
def method_vectorized():
    start_time = time.time()
    distances = haversine(ref_lat, ref_long,
                         df['locLat'].values,
                         df['locLong'].values)
    end_time = time.time()
    return distances, end_time - start_time

# For loop method
distances_loop, time_loop = method_loop()
print(f"For loop time: {time_loop:.4f} seconds")

# Apply method
distances_apply, time_apply = method_apply()
print(f"Apply method time: {time_apply:.4f} seconds")

# Vectorized method
distances_vector, time_vector = method_vectorized()
print(f"Vectorized time: {time_vector:.4f} seconds")

File: week03/Venkatesh Yadav Edamoni WEEK03.R
library(tidyverse)
library(microbenchmark)

data <- read_excel("clinics.xls")

data <- data %>%
  mutate(
    locLat = as.numeric(locLat),
    locLong = as.numeric(locLong)
  )

haversine <- function(lat1, lon1, lat2, lon2) {
  lat1 <- lat1 * pi / 180
  lon1 <- lon1 * pi / 180
  lat2 <- lat2 * pi / 180
  lon2 <- lon2 * pi / 180
  dlat <- lat2 - lat1
  dlon <- lon2 - lon1
  a <- sin(dlat/2)^2 + cos(lat1) * cos(lat2) * sin(dlon/2)^2
  c <- 2 * asin(sqrt(a))
  r <- 3959
  
  return(r * c)
}

# Reference point (first clinic)
ref_lat <- data$locLat[1]
ref_long <- data$locLong[1]

# Method 1: For loop
method_loop <- function() {
  distances <- numeric(nrow(data))
  for(i in 1:nrow(data)) {
    distances[i] <- haversine(ref_lat, ref_long,
                              data$locLat[i],
                              data$locLong[i])
  }
  return(distances)
}

# Method 2: Using apply
method_apply <- function() {
  apply(data, 1, function(row) {
    haversine(ref_lat, ref_long,
              as.numeric(row['locLat']),
              as.numeric(row['locLong']))
  })
}

# Method 3: Vectorized
method_vectorized <- function() {
  haversine(ref_lat, ref_long,
            data$locLat,
            data$locLong)
}

cat("Running benchmarks...\n")
results <- microbenchmark(
  loop = method_loop(),
  apply = method_apply(),
  vectorized = method_vectorized(),
  times = 100
)

print(results)
